tests/roundrobin_poll_test.cpp

```diff
@@ -0,0 +1,162 @@
+    int messages_per_partition = 3;
+    int total_messages = KAFKA_NUM_PARTITIONS * messages_per_partition;
+
+    // Create a consumer and subscribe to the topic
+    Consumer consumer(make_consumer_config());
+    TopicPartitionList partitions;
+    for (int i = 0; i < KAFKA_NUM_PARTITIONS; partitions.emplace_back(KAFKA_TOPICS[0], i++));
+    consumer.assign(partitions);
+    
+    // Start the runner with the original consumer
+    ConsumerRunner runner(consumer, total_messages, KAFKA_NUM_PARTITIONS);
+
+    // Produce messages so we stop the consumer
+    Producer producer(make_producer_config());
+    string payload = "Serial";
+    
+    // push 3 messages in each partition
+    for (int i = 0; i < total_messages; ++i) {
+        producer.produce(MessageBuilder(KAFKA_TOPICS[0]).partition(i%KAFKA_NUM_PARTITIONS).payload(payload));
+    }
+    producer.flush();
+    runner.try_join();
+    
+    // Check that we have all messages
+    REQUIRE(runner.get_messages().size() == total_messages);
+    
+    // messages should have sequential identical partition ids in groups of <messages_per_partition>
+    int expected_partition;
+    for (int i = 0; i < total_messages; ++i) {
+        if ((i % messages_per_partition) == 0) {
+            expected_partition = runner.get_messages()[i].get_partition();
+        }
+        REQUIRE(runner.get_messages()[i].get_partition() == expected_partition);
+        REQUIRE((string)runner.get_messages()[i].get_payload() == payload);
+    }
+}
+
+TEST_CASE("roundrobin consumer test", "[roundrobin consumer]") {
+    TopicPartitionList assignment;
+    int messages_per_partition = 3;
+    int total_messages = KAFKA_NUM_PARTITIONS * messages_per_partition;
+
+    // Create a consumer and subscribe to the topic
+    PollStrategyAdapter consumer(make_consumer_config());
+    TopicPartitionList partitions;
+    for (int i = 0; i < KAFKA_NUM_PARTITIONS; partitions.emplace_back(KAFKA_TOPICS[0], i++));
+    consumer.assign(partitions);
+    consumer.add_polling_strategy(unique_ptr<PollInterface>(new RoundRobinPollStrategy(consumer)));
+    
+    PollConsumerRunner runner(consumer, total_messages, KAFKA_NUM_PARTITIONS);
+
+    // Produce messages so we stop the consumer
+    Producer producer(make_producer_config());
+    string payload = "RoundRobin";
+    
+    // push 3 messages in each partition
+    for (int i = 0; i < total_messages; ++i) {
+        producer.produce(MessageBuilder(KAFKA_TOPICS[0]).partition(i%KAFKA_NUM_PARTITIONS).payload(payload));
+    }
+    producer.flush();
+    runner.try_join();
+    
+    // Check that we have all messages
+    REQUIRE(runner.get_messages().size() == total_messages);
+    
+    // Check that we have one message from each partition in desired order
+    vector<int> partition_order = make_roundrobin_partition_vector(total_messages);
+    
+    for (int i = 0; i < total_messages; ++i) {
+        REQUIRE(runner.get_messages()[i].get_partition() == partition_order[i+1]);
```

Comment:
I'm probably missing something but how does this guarantee the starting topic/partition will be what you expect? If say, you somehow took slightly longer to produce messages than you expect, you could end up having the consumer poll through the first partition, jump to the second and then get a message on that one. The tests seem to have succeeded but I'm curious about that potential race condition.

Reply:
Well most likely the partitions are emtpy, which means all the EOFs arrive in sequence, and then the first poll timeout is 1s which is more then enough to publish 3 messages, especially since kafka broker is on localhost. In my testing, the broker is in a remote datacenter cluster and it still works. I have slightly changed the test now, allowing for a different start partition in case the first poll (or other ones) times out.